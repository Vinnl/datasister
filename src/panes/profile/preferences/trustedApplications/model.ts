export type Mode = 'Read' | 'Append' | 'Write' | 'Control'

export interface TrustedApplication {
  origin: string
  subject: string
  modes: Mode[]
}
