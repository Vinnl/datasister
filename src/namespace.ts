import $rdf from 'rdflib';

const aliases: {[alias: string]: string} = {
  acl: 'http://www.w3.org/ns/auth/acl#',
  arg: 'http://www.w3.org/ns/pim/arg#',
  cal: 'http://www.w3.org/2002/12/cal/ical#',
  contact: 'http://www.w3.org/2000/10/swap/pim/contact#',
  dc: 'http://purl.org/dc/elements/1.1/',
  dct: 'http://purl.org/dc/terms/',
  doap: 'http://usefulinc.com/ns/doap#',
  foaf: 'http://xmlns.com/foaf/0.1/',
  http: 'http://www.w3.org/2007/ont/http#',
  httph: 'http://www.w3.org/2007/ont/httph#',
  icalTZ: 'http://www.w3.org/2002/12/cal/icaltzd#', // Beware: not cal:
  ldp: 'http://www.w3.org/ns/ldp#',
  link: 'http://www.w3.org/2007/ont/link#',
  log: 'http://www.w3.org/2000/10/swap/log#',
  meeting: 'http://www.w3.org/ns/pim/meeting#',
  mo: 'http://purl.org/ontology/mo/',
  owl: 'http://www.w3.org/2002/07/owl#',
  pad: 'http://www.w3.org/ns/pim/pad#',
  patch: 'http://www.w3.org/ns/pim/patch#',
  qu: 'http://www.w3.org/2000/10/swap/pim/qif#',
  trip: 'http://www.w3.org/ns/pim/trip#',
  rdf: 'http://www.w3.org/1999/02/22-rdf-syntax-ns#',
  rdfs: 'http://www.w3.org/2000/01/rdf-schema#',
  rss: 'http://purl.org/rss/1.0/',
  sched: 'http://www.w3.org/ns/pim/schedule#',
  schema: 'http:/schema.org/', // @@ beware confusion with documents no 303
  sioc: 'http://rdfs.org/sioc/ns#',
  solid: 'http://www.w3.org/ns/solid/terms#',
  space: 'http://www.w3.org/ns/pim/space#',
  stat: 'http://www.w3.org/ns/posix/stat#',
  tab: 'http://www.w3.org/2007/ont/link#',
  tabont: 'http://www.w3.org/2007/ont/link#',
  ui: 'http://www.w3.org/ns/ui#',
  vcard: 'http://www.w3.org/2006/vcard/ns#',
  wf: 'http://www.w3.org/2005/01/wf/flow#',
  xsd: 'http://www.w3.org/2001/XMLSchema#'
}

type Namespace = (localName: string) => $rdf.NamedNode;
type Namespaces = {[alias: string]: Namespace};
export const namespaces: Namespaces = Object.keys(aliases).reduce(
  (namespaces, alias) => {
    namespaces[alias] = (localName: string) => $rdf.namedNode(aliases[alias] + localName);
    return namespaces;
  },
  {} as Namespaces,
);
