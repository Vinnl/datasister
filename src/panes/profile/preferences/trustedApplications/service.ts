import $rdf, { NamedNode, IndexedFormula, Statement } from 'rdflib'
import { namespaces as ns } from '../../../../namespace';
import { Mode, TrustedApplication } from './model'

export function getStatementsToDelete (
  origin: NamedNode,
  person: NamedNode,
  store: IndexedFormula,
) {
  const applicationStatements = store.statementsMatching(null, ns.acl('origin'), origin, null)
  const statementsToDelete = applicationStatements.reduce(
    (memo, st) => {
      return memo
        .concat(store.statementsMatching(person, ns.acl('trustedApp'), st.subject, null, false))
        .concat(store.statementsMatching(st.subject, null, null, null, false))
    },
    [] as Statement[]
  )
  return statementsToDelete
}

export function getStatementsToAdd (
  origin: NamedNode,
  nodeName: string,
  modes: Mode[],
  person: NamedNode,
) {
  var application = new $rdf.BlankNode(`bn_${nodeName}`)
  return [
    $rdf.st(person, ns.acl('trustedApp'), application, person.doc()),
    $rdf.st(application, ns.acl('origin'), origin, person.doc()),
    ...modes
      .map(mode => {
        return ns.acl(mode)
      })
      .map(mode => $rdf.st(application, ns.acl('mode'), mode, person.doc()))
  ]
}

/* istanbul ignore next [This executes the actual HTTP requests, which is too much effort to test.] */
export function fetchTrustedApps (
  store: $rdf.IndexedFormula,
  subject: $rdf.NamedNode,
): TrustedApplication[] {
  return (store.each(subject, ns.acl('trustedApp'), undefined, undefined) as any)
    .flatMap((app: $rdf.NamedNode) => {
      return store.each(app, ns.acl('origin'), undefined, undefined)
        .map((origin) => {
          const modes = store.each(app, ns.acl('mode'), undefined, undefined)
          const trustedApp: TrustedApplication = {
            origin: origin.value,
            subject: subject.value,
            modes: modes.map((mode) => deserialiseMode(mode as $rdf.NamedNode))
          }
          return trustedApp
        })
    })
    .sort((appA: TrustedApplication, appB: TrustedApplication) => (appA.origin > appB.origin) ? 1 : -1)
}

/**
 * @param serialisedMode The full IRI of a mode
 * @returns A plain text string representing that mode, i.e. 'Read', 'Append', 'Write' or 'Control'
 */
export function deserialiseMode (serialisedMode: $rdf.NamedNode): Mode {
  const deserialisedMode = serialisedMode.value
    .replace(ns.acl('Read').value, 'Read')
    .replace(ns.acl('Append').value, 'Append')
    .replace(ns.acl('Write').value, 'Write')
    .replace(ns.acl('Control').value, 'Control')

  return deserialisedMode as Mode
}
