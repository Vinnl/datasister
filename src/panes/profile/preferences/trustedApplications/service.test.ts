/* eslint-env jest */
import $rdf from 'rdflib';
import { getStatementsToDelete, getStatementsToAdd, deserialiseMode } from './service';
import { namespaces as ns } from '../../../../namespace';
import { Mode } from './model';

describe('getStatementsToDelete', () => {
  it('should return an empty array when there are no statements', () => {
    const mockStore = $rdf.graph()
    const mockOrigin = $rdf.sym('https://fake.origin')
    const mockProfile = $rdf.sym('https://fake.profile#me')
    expect(getStatementsToDelete(mockOrigin, mockProfile, mockStore)).toEqual([])
  })

  it('should return all statements for the given origin', () => {
    const mockStore = $rdf.graph();
    const mockApplication = $rdf.sym('https://fake.app');
    const mockOrigin = $rdf.sym('https://fake.origin');
    const mockProfile = $rdf.sym('https://fake.profile#me');
    mockStore.add(mockApplication, ns.acl('origin'), mockOrigin, mockStore);
    mockStore.add(mockApplication, ns.acl('mode'), ns.acl('Read'), mockStore);
    mockStore.add(mockProfile, ns.acl('trustedApp'), mockApplication, mockStore);
    const statementsToDelete = getStatementsToDelete(mockOrigin, mockProfile, mockStore);
    expect(statementsToDelete.length).toBe(3);
    expect(statementsToDelete).toMatchSnapshot();
  });

  it('should not return statements for a different origin', () => {
    const mockStore = $rdf.graph();
    const mockApplication = $rdf.sym('https://fake.app');
    const mockOrigin = $rdf.sym('https://fake.origin');
    const mockProfile = $rdf.sym('https://fake.profile#me');
    mockStore.add(mockApplication, ns.acl('origin'), mockOrigin, mockStore);
    mockStore.add(mockApplication, ns.acl('mode'), ns.acl('Read'), mockStore);
    mockStore.add(mockProfile, ns.acl('trustedApp'), mockApplication, mockStore);

    const statementsToDelete = getStatementsToDelete(($rdf.lit as any)('A different origin'), mockProfile, mockStore);
    expect(statementsToDelete.length).toBe(0);
    expect(statementsToDelete).toEqual([]);
  });
});

describe('getStatementsToAdd', () => {
  it('should return all required statements to add the given permissions for a given origin', () => {
    const mockOrigin = $rdf.sym('https://fake.origin');
    const mockProfile = $rdf.sym('https://fake.profile#me');
    const modes: Mode[] = ['read', 'write'];

    const statementsToAdd = getStatementsToAdd(mockOrigin, 'mock_app_id', modes, mockProfile);
    expect(statementsToAdd.length).toBe(4);
    expect(statementsToAdd).toMatchSnapshot();
  });
});

describe('deserialiseMode', () => {
  it('should convert a full namespaced ACL to a plaintext string', () => {
    expect(deserialiseMode(ns.acl('Read'))).toBe('Read');
    expect(deserialiseMode(ns.acl('Append'))).toBe('Append');
    expect(deserialiseMode(ns.acl('Write'))).toBe('Write');
    expect(deserialiseMode(ns.acl('Control'))).toBe('Control');
  });
});
