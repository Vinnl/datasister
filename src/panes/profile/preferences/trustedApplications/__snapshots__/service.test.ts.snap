// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getStatementsToAdd should return all required statements to add the given permissions for a given origin 1`] = `
Array [
  Statement {
    "object": BlankNode {
      "id": "bn_mock_app_id",
      "termType": "BlankNode",
      "value": "bn_mock_app_id",
    },
    "predicate": NamedNode {
      "termType": "NamedNode",
      "value": "http://www.w3.org/ns/auth/acl#trustedApp",
    },
    "subject": NamedNode {
      "termType": "NamedNode",
      "value": "https://fake.profile#me",
    },
    "why": NamedNode {
      "termType": "NamedNode",
      "value": "https://fake.profile",
    },
  },
  Statement {
    "object": NamedNode {
      "termType": "NamedNode",
      "value": "https://fake.origin",
    },
    "predicate": NamedNode {
      "termType": "NamedNode",
      "value": "http://www.w3.org/ns/auth/acl#origin",
    },
    "subject": BlankNode {
      "id": "bn_mock_app_id",
      "termType": "BlankNode",
      "value": "bn_mock_app_id",
    },
    "why": NamedNode {
      "termType": "NamedNode",
      "value": "https://fake.profile",
    },
  },
  Statement {
    "object": NamedNode {
      "termType": "NamedNode",
      "value": "http://www.w3.org/ns/auth/acl#read",
    },
    "predicate": NamedNode {
      "termType": "NamedNode",
      "value": "http://www.w3.org/ns/auth/acl#mode",
    },
    "subject": BlankNode {
      "id": "bn_mock_app_id",
      "termType": "BlankNode",
      "value": "bn_mock_app_id",
    },
    "why": NamedNode {
      "termType": "NamedNode",
      "value": "https://fake.profile",
    },
  },
  Statement {
    "object": NamedNode {
      "termType": "NamedNode",
      "value": "http://www.w3.org/ns/auth/acl#write",
    },
    "predicate": NamedNode {
      "termType": "NamedNode",
      "value": "http://www.w3.org/ns/auth/acl#mode",
    },
    "subject": BlankNode {
      "id": "bn_mock_app_id",
      "termType": "BlankNode",
      "value": "bn_mock_app_id",
    },
    "why": NamedNode {
      "termType": "NamedNode",
      "value": "https://fake.profile",
    },
  },
]
`;

exports[`getStatementsToDelete should return all statements for the given origin 1`] = `
Array [
  Statement {
    "object": NamedNode {
      "termType": "NamedNode",
      "value": "https://fake.app",
    },
    "predicate": NamedNode {
      "termType": "NamedNode",
      "value": "http://www.w3.org/ns/auth/acl#trustedApp",
    },
    "subject": NamedNode {
      "termType": "NamedNode",
      "value": "https://fake.profile#me",
    },
    "why": IndexedFormula {
      "HTTPRedirects": Array [],
      "aliases": Array [],
      "classActions": Array [],
      "constraints": Array [],
      "features": Array [
        "sameAs",
        "InverseFunctionalProperty",
        "FunctionalProperty",
      ],
      "index": Array [
        Array [],
        Array [],
        Array [],
        Array [],
      ],
      "initBindings": Array [],
      "namespaces": Object {},
      "objectIndex": Array [],
      "optional": Array [],
      "predicateIndex": Array [],
      "propertyActions": Array [],
      "redirections": Array [],
      "statements": Array [
        Statement {
          "object": NamedNode {
            "termType": "NamedNode",
            "value": "https://fake.origin",
          },
          "predicate": NamedNode {
            "termType": "NamedNode",
            "value": "http://www.w3.org/ns/auth/acl#origin",
          },
          "subject": NamedNode {
            "termType": "NamedNode",
            "value": "https://fake.app",
          },
          "why": [Circular],
        },
        Statement {
          "object": NamedNode {
            "termType": "NamedNode",
            "value": "http://www.w3.org/ns/auth/acl#Read",
          },
          "predicate": NamedNode {
            "termType": "NamedNode",
            "value": "http://www.w3.org/ns/auth/acl#mode",
          },
          "subject": NamedNode {
            "termType": "NamedNode",
            "value": "https://fake.app",
          },
          "why": [Circular],
        },
        [Circular],
      ],
      "subjectIndex": Array [],
      "termType": "Graph",
      "whyIndex": Array [],
    },
  },
  Statement {
    "object": NamedNode {
      "termType": "NamedNode",
      "value": "https://fake.origin",
    },
    "predicate": NamedNode {
      "termType": "NamedNode",
      "value": "http://www.w3.org/ns/auth/acl#origin",
    },
    "subject": NamedNode {
      "termType": "NamedNode",
      "value": "https://fake.app",
    },
    "why": IndexedFormula {
      "HTTPRedirects": Array [],
      "aliases": Array [],
      "classActions": Array [],
      "constraints": Array [],
      "features": Array [
        "sameAs",
        "InverseFunctionalProperty",
        "FunctionalProperty",
      ],
      "index": Array [
        Array [],
        Array [],
        Array [],
        Array [],
      ],
      "initBindings": Array [],
      "namespaces": Object {},
      "objectIndex": Array [],
      "optional": Array [],
      "predicateIndex": Array [],
      "propertyActions": Array [],
      "redirections": Array [],
      "statements": Array [
        [Circular],
        Statement {
          "object": NamedNode {
            "termType": "NamedNode",
            "value": "http://www.w3.org/ns/auth/acl#Read",
          },
          "predicate": NamedNode {
            "termType": "NamedNode",
            "value": "http://www.w3.org/ns/auth/acl#mode",
          },
          "subject": NamedNode {
            "termType": "NamedNode",
            "value": "https://fake.app",
          },
          "why": [Circular],
        },
        Statement {
          "object": NamedNode {
            "termType": "NamedNode",
            "value": "https://fake.app",
          },
          "predicate": NamedNode {
            "termType": "NamedNode",
            "value": "http://www.w3.org/ns/auth/acl#trustedApp",
          },
          "subject": NamedNode {
            "termType": "NamedNode",
            "value": "https://fake.profile#me",
          },
          "why": [Circular],
        },
      ],
      "subjectIndex": Array [],
      "termType": "Graph",
      "whyIndex": Array [],
    },
  },
  Statement {
    "object": NamedNode {
      "termType": "NamedNode",
      "value": "http://www.w3.org/ns/auth/acl#Read",
    },
    "predicate": NamedNode {
      "termType": "NamedNode",
      "value": "http://www.w3.org/ns/auth/acl#mode",
    },
    "subject": NamedNode {
      "termType": "NamedNode",
      "value": "https://fake.app",
    },
    "why": IndexedFormula {
      "HTTPRedirects": Array [],
      "aliases": Array [],
      "classActions": Array [],
      "constraints": Array [],
      "features": Array [
        "sameAs",
        "InverseFunctionalProperty",
        "FunctionalProperty",
      ],
      "index": Array [
        Array [],
        Array [],
        Array [],
        Array [],
      ],
      "initBindings": Array [],
      "namespaces": Object {},
      "objectIndex": Array [],
      "optional": Array [],
      "predicateIndex": Array [],
      "propertyActions": Array [],
      "redirections": Array [],
      "statements": Array [
        Statement {
          "object": NamedNode {
            "termType": "NamedNode",
            "value": "https://fake.origin",
          },
          "predicate": NamedNode {
            "termType": "NamedNode",
            "value": "http://www.w3.org/ns/auth/acl#origin",
          },
          "subject": NamedNode {
            "termType": "NamedNode",
            "value": "https://fake.app",
          },
          "why": [Circular],
        },
        [Circular],
        Statement {
          "object": NamedNode {
            "termType": "NamedNode",
            "value": "https://fake.app",
          },
          "predicate": NamedNode {
            "termType": "NamedNode",
            "value": "http://www.w3.org/ns/auth/acl#trustedApp",
          },
          "subject": NamedNode {
            "termType": "NamedNode",
            "value": "https://fake.profile#me",
          },
          "why": [Circular],
        },
      ],
      "subjectIndex": Array [],
      "termType": "Graph",
      "whyIndex": Array [],
    },
  },
]
`;
